src/devices/src/legacy/serial.rs

```diff
@@ -244,10 +240,26 @@ impl Serial {
         }
     }
 
+    fn recv_bytes(&mut self) -> io::Result<usize> {
+        if let Some(input) = self.input.as_mut() {
+            let mut out = [0u8; 32];
+            return match input.read(&mut out) {
+                Ok(count) if count > 0 => {
+                    self.raw_input(&out[..count])?;
+                    Ok(count)
+                }
+                Err(e) => Err(e),
+                _ => Ok(0),
+            };
```

Comment:
nit that gets rid of the strange `Err(e) => Err(e)`
```suggestion
            input.read(&mut out).and_then(|count| {
                if count > 0 {
                    self.raw_input(&out[..count])?;
                    Ok(count)
                } else {
                    Ok(0)
                }
            })?;
```